File: tools/studio/src/panels/agents.js
// SPDX-License-Identifier: MIT
//
// Agents: the CrewAI crew's surface in the studio. A person writes a brief;
// the studio creates the worktree and hands the flow to the agents container;
// the flow patches its record as it runs. A task awaiting a human shows
// approve / reject; an approved, finished one shows merge (fast-forward or
// refuse) — the studio merges, never the agent.

import { el, get, post } from "../api.js";

const fmtTime = (t) => (t ? new Date(t * 1000).toLocaleString() : "");

export async function render(root, params) {
  return params[0] ? detail(root, params[0]) : list(root);
}

async function list(root) {
  const brief = el("textarea", { rows: 3, placeholder: "a task for the crew — e.g. “add a pillar to assets/map_demo.map”", style: "width:100%" });
  const status = el("div", { class: "dim", text: "" });
  const submit = el("button", {
    onclick: async () => {
      submit.disabled = true;
      try {
        const task = await post("/api/agent-tasks", { brief: brief.value });
        location.hash = `#/agents/${task.id}`;
      } catch (e) {
        status.textContent = e.message;
        submit.disabled = false;
      }
    },
  }, "start task");

  const draw = async () => {
    const { tasks, agents_service } = await get("/api/agent-tasks");
    listBody.replaceChildren(
      agents_service ? null : el("p", { class: "warn", text: "no agents service on the network (KILN_AGENTS_URL unset) — tasks queue as worktrees until the agents container is up." }),
      tasks.length === 0 ? el("p", { class: "dim", text: "No tasks yet." }) :
        el("table", {},
          el("tr", {}, ["when", "task", "who", "phase", "decision"].map((h) => el("th", { text: h }))),
          tasks.map((t) => el("tr", {},
            el("td", { text: fmtTime(t.created) }),
            el("td", {}, el("a", { href: `#/agents/${t.id}`, text: t.brief.slice(0, 80) })),
            el("td", { text: t.by }),
            el("td", { class: `state-${t.phase === "done" ? "ok" : "running"}`, text: t.phase || "" }),
            el("td", { text: t.approved === true ? `approved by ${t.decided_by}` :
                             t.approved === false ? `rejected by ${t.decided_by}` : "—" })))));
  };
  const listBody = el("div", {});
  root.replaceChildren(el("h1", { text: "Agent tasks" }), brief, submit, status, listBody);
  await draw();
  const t = setInterval(() => draw().catch(() => {}), 3000);
  return () => clearInterval(t);
}

async function detail(root, id) {
  const errBox = el("p", { class: "err", text: "" });
  const err = (msg) => { errBox.textContent = msg; };
  const draw = async () => {
    const t = await get(`/api/agent-tasks/${id}`);
    const actions = el("div", { class: "chips" });
    if (t.phase === "awaiting-human" && t.approved === null) {
      actions.replaceChildren(
        el("button", { onclick: () => post(`/api/agent-tasks/${id}/approve`).catch((e) => err(e.message)) }, "approve"),
        el("button", { onclick: () => post(`/api/agent-tasks/${id}/reject`).catch((e) => err(e.message)) }, "reject"));
    } else if (t.phase === "done" && t.approved === true && !t.merged) {
      actions.replaceChildren(
        el("button", { onclick: () => post(`/api/agent-tasks/${id}/merge`).catch((e) => err(e.message)) }, "merge (fast-forward)"),
        el("button", { onclick: () => post(`/api/agent-tasks/${id}/discard`).catch((e) => err(e.message)) }, "discard"));
    } else if (["queued", "running"].includes(t.phase) || t.approved === false) {
      actions.replaceChildren(
        el("button", { onclick: () => post(`/api/agent-tasks/${id}/discard`).catch((e) => err(e.message)) }, "discard"));
    } else {
      actions.replaceChildren();
    }
    errBox.textContent = t.note || "";
    body.replaceChildren(
      el("div", { class: "row" }, el("h1", { text: `task ${t.id}` }),
         el("span", { class: `state-${t.phase === "done" || t.phase === "merged" ? "ok" : "running"}`, text: t.phase })),
      el("p", {}, el("b", { text: t.brief }), ` — ${t.by}, ${fmtTime(t.created)}`),
      t.kind ? el("p", { class: "dim", text: `kind: ${t.kind} · branch ${t.branch}` }) : el("p", { class: "dim", text: `branch ${t.branch}` }),
      actions,
      t.plan ? el("h2", { text: "plan" }) : null,
      t.plan ? el("pre", { class: "log", text: t.plan }) : null,
      t.validations ? el("h2", { text: "validations" }) : null,
      t.validations ? el("pre", { class: "log", text: JSON.stringify(t.validations, null, 2) }) : null,
      t.review ? el("h2", { text: "review" }) : null,
      t.review ? el("pre", { class: "log", text: t.review }) : null,
      t.diff_stat ? el("h2", { text: "diff" }) : null,
      t.diff_stat ? el("pre", { class: "log", text: t.diff_stat + (t.ahead ? "\n\n" + t.ahead : "") }) : null,
      t.report ? el("h2", { text: "final report" }) : null,
      t.report ? el("pre", { class: "log", text: JSON.stringify(t.report, null, 2) }) : null);
  };
  const body = el("div", {});
  root.replaceChildren(el("a", { href: "#/agents", text: "← all tasks" }), errBox, body);
  await draw();
  const t = setInterval(() => draw().catch(() => {}), 3000);
  return () => clearInterval(t);
}

File: tools/studio/studio/agents.py
# SPDX-License-Identifier: MIT
"""agents.py — the studio side of the agent loop: tasks, worktrees, merges.

The STUDIO owns the worktree lifecycle, not the agent (a runtime that can
create its own worktrees can lose them; a studio that can creates them, shows
their diff, and merges or deletes). One task = one worktree at
``.studio/agents/worktrees/<task_id>`` on branch ``agent/<task_id>``, a record in
``.studio/agents/tasks.json``, and a flow run in the agents container kicked
over HTTP (KILN_AGENTS_URL).

Records keep: brief, who asked (a human via the panel, or the mesh later),
phase (patched by the flow as it runs), the plan/review/validations the flow
reports, the human's decision, and the eventual merge outcome. The flow's
PATCH is whitelisted — it can report, never decide; `approved` moves only
through decide(), called from the approve/reject endpoint a person clicks.

Merge is fast-forward-or-refuse: the flow's branch must contain the served
checkout's HEAD exactly. A branch that has drifted behind work the humans
did is refused, never silently merged.
"""

import json
import os
import subprocess
import threading
import time
import urllib.request
from pathlib import Path

PATCHABLE = {"phase", "plan", "kind", "validations", "review", "merge_request", "report"}


class AgentTasks:
    def __init__(self, repo: Path):
        self.repo = repo
        self.dir = repo / ".studio" / "agents"
        self.dir.mkdir(parents=True, exist_ok=True)
        self.file = self.dir / "tasks.json"
        self.lock = threading.Lock()
        self.tasks = self._load()
        self.url = os.environ.get("KILN_AGENTS_URL", "").rstrip("/")

    # ── storage ───────────────────────────────────────────────────────────
    def _load(self) -> dict:
        if self.file.is_file():
            try:
                return json.loads(self.file.read_text())
            except (OSError, json.JSONDecodeError):
                pass
        return {}

    def _save(self):
        tmp = self.file.with_suffix(".tmp")
        tmp.write_text(json.dumps(self.tasks, indent=1) + "\n")
        os.replace(tmp, self.file)

    def _git(self, *argv, cwd=None, check=True):
        proc = subprocess.run(["git", *argv], cwd=cwd or self.repo,
                              capture_output=True, text=True, timeout=120)
        if check and proc.returncode != 0:
            raise ValueError(proc.stderr.strip() or f"git {' '.join(argv)} failed")
        return proc.stdout.strip()

    def _call_agents(self, method: str, path: str, body: dict | None = None):
        if not self.url:
            raise ValueError("no agents service configured (KILN_AGENTS_URL unset)")
        req = urllib.request.Request(
            self.url + path, method=method,
            data=json.dumps(body).encode() if body is not None else None,
            headers={"Content-Type": "application/json"})
        with urllib.request.urlopen(req, timeout=15) as r:
            return json.loads(r.read() or b"{}")

    # ── the lifecycle ─────────────────────────────────────────────────────
    def create(self, brief: str, by: str) -> dict:
        brief = brief.strip()
        if not brief:
            raise ValueError("a task needs a brief")
        task_id = f"t{int(time.time()) % 10_000_000:07d}-{os.urandom(3).hex()}"
        branch = f"agent/{task_id}"
        worktree = self.dir / "worktrees" / task_id
        worktree.parent.mkdir(parents=True, exist_ok=True)
        self._git("worktree", "add", str(worktree), "-b", branch)
        rec = {"id": task_id, "brief": brief[:2000], "by": by, "branch": branch,
               "worktree": str(worktree), "phase": "created", "approved": None,
               "created": int(time.time())}
        try:
            self._call_agents("POST", "/task",
                              {"task_id": task_id, "brief": brief, "worktree": str(worktree)})
            rec["phase"] = "running"
        except (ValueError, OSError) as e:
            rec["phase"] = "queued"
            rec["note"] = f"agents service not reached ({e}); the task and its worktree exist — " \
                          "start the agents container and PATCH it, or delete and retry"
        with self.lock:
            self.tasks[task_id] = {k: v for k, v in rec.items()}
            self._save()
        return rec

    def patch(self, task_id: str, fields: dict) -> dict:
        """The flow reporting in. Whitelisted keys only — a flow reports, it
        never decides (approved) and never deletes (worktree)."""
        with self.lock:
            rec = self.tasks.get(task_id)
            if rec is None:
                raise KeyError(task_id)
            for k, v in fields.items():
                if k in PATCHABLE:
                    rec[k] = v
            self._save()
            return rec

    def decide(self, task_id: str, approved: bool, by: str) -> dict:
        """A person's yes/no. Recorded, then relayed to the agents service,
        which re-kicks the persisted flow so await_human → finish completes."""
        with self.lock:
            rec = self.tasks.get(task_id)
            if rec is None:
                raise KeyError(task_id)
            if rec.get("approved") is not None:
                raise ValueError("already decided")
            rec["approved"] = bool(approved)
            rec["decided_by"] = by
            self._save()
        try:
            self._call_agents("POST", f"/task/{task_id}/decision", {"approved": approved})
        except (ValueError, OSError) as e:
            with self.lock:
                rec["note"] = f"decision recorded; the agents service was not reached ({e}) — " \
                              "the flow resumes when it is"
                self._save()
        return rec

    def merge(self, task_id: str) -> dict:
        """Fast-forward the served checkout to the agent branch, or refuse.
        On success the worktree goes away and the branch is deleted — a
        merged task leaves no residue."""
        with self.lock:
            rec = self.tasks.get(task_id)
            if rec is None:
                raise KeyError(task_id)
            if rec.get("approved") is not True:
                raise ValueError("merge needs an approval first")
            if rec.get("phase") not in ("done",):
                raise ValueError(f"the flow is not done (phase: {rec.get('phase')})")
        branch = rec["branch"]
        self._git("merge", "--ff-only", branch)  # raises when not a fast-forward
        self._git("worktree", "remove", "--force", rec["worktree"], check=False)
        self._git("branch", "-D", branch, check=False)
        with self.lock:
            rec["merged"] = True
            rec["phase"] = "merged"
            self._save()
        return rec

    def discard(self, task_id: str) -> dict:
        with self.lock:
            rec = self.tasks.get(task_id)
            if rec is None:
                raise KeyError(task_id)
        self._git("worktree", "remove", "--force", rec["worktree"], check=False)
        self._git("branch", "-D", rec["branch"], check=False)
        with self.lock:
            rec["discarded"] = True
            rec["phase"] = "discarded"
            self._save()
        return rec

    def get(self, task_id: str, with_diff=True) -> dict | None:
        with self.lock:
            rec = dict(self.tasks.get(task_id) or {})
        if not rec:
            return None
        if with_diff and Path(rec.get("worktree", "/nonexistent")).is_dir():
            rec["diff_stat"] = self._git("diff", "--stat", "HEAD", check=False,
                                         cwd=rec["worktree"])
            base = self._git("rev-parse", "HEAD", check=False)
            if base:
                rec["ahead"] = self._git("log", "--oneline", f"{base}..HEAD",
                                         check=False, cwd=rec["worktree"])
        rec["agents_service"] = bool(self.url)
        return rec

    def list(self) -> list:
        with self.lock:
            return sorted((dict(r) for r in self.tasks.values()),
                          key=lambda r: r.get("created", 0), reverse=True)
